File: DragonsGenerator.WEB/src/app/core/models/Species/species.ts
// src/app/core/models/Species/species.ts

// ASP.NET sérialise par défaut en camelCase, donc on utilise camelCase ici.

export interface Source {
  book: string;
  pages: string;
}

export interface Flavor {
  summary: string;
  culture?: string;
  origins?: string;
  loreNotes?: string[];
}

export interface Height {
  desc: string;
  rangeM?: string;
}

export interface Weight {
  desc: string;
  rangeKg?: string;
}

export interface Age {
  maturityYears: number;
  lifespanYears: number;
  desc: string;
  adulthoodCulturalYears?: number;
  lifespanMaxYears?: number;
}

export interface Alignment {
  tendency: string;
  desc: string;
}

export interface FlexibleAsi {
  count: number;
  value: number;
  excluded: string[];
}

export interface BaseStats {
  abilityScoreIncrease: Record<string, number>;
  speedM: number;
  size: string;
  darkvisionM: number;
  height: Height;
  weight: Weight;
  age: Age;
  alignment: Alignment;
  flexibleAsi?: FlexibleAsi;
  speedNotes?: string;
  speedNotReducedByHeavyArmor?: boolean;
}

// `mechanics` est un JsonElement côté C# avec des formes très variables
// selon le trait/règle. On le type en `unknown` (à narrower au cas par cas
// dans les composants si besoin) plutôt qu'en `any`.
export interface Trait {
  id: string;
  name: string;
  desc: string;
  mechanics?: unknown;
}

export interface CreationChoice {
  id: string;
  name: string;
  desc: string;
  type: string;
  choiceCount?: number;
  options?: unknown; // tableau hétérogène d'options
  optionGroups?: unknown; // groupes d'options (cf. melessë)
  spellList?: string;
  spellLevel?: number;
  spellcastingAbility?: string;
  valuePerChoice?: number;
  excluded?: string[];
}

export interface Languages {
  fixed: string[];
  choiceCount: number;
  notes?: string;
  grantsFromChoice?: unknown;
}

export interface Subspecies {
  id: string;
  name: string;
  playable: boolean;
  flavor: string;
  abilityScoreIncrease: Record<string, number>;
  traits: Trait[];
  creationChoices: CreationChoice[];
  playableNotes?: string;
  languages?: Languages;
}

export interface OptionalRule {
  id: string;
  name: string;
  desc: string;
  mechanics?: unknown;
}

export interface CivilizationLink {
  id: string;
  name: string;
  desc: string;
}

export interface Species {
  id: string;
  name: string;
  nameAlt: string[];
  source: Source;
  flavor: Flavor;
  baseStats: BaseStats;
  traits: Trait[];
  creationChoices: CreationChoice[];
  languages: Languages;
  subspecies: Subspecies[];
  optionalRules: OptionalRule[];
  civilizationLinks?: CivilizationLink[];
}

File: DragonsGenerator.API/Models/Species.cs
using System.Text.Json;

namespace DragonsGenerator.API.Models;

// Racine du fichier JSON
public record SpeciesData(
    Dictionary<string, string> SizeCodes,
    Dictionary<string, string> AbilityCodes,
    List<Species> Species
);

public record Species(
    string Id,
    string Name,
    List<string> NameAlt,
    Source Source,
    Flavor Flavor,
    BaseStats BaseStats,
    List<Trait> Traits,
    List<CreationChoice> CreationChoices,
    Languages Languages,
    List<Subspecies> Subspecies,
    List<OptionalRule> OptionalRules,
    List<CivilizationLink>? CivilizationLinks = null
);

public record Source(string Book, string Pages);

public record Flavor(
    string Summary,
    string? Culture = null,
    string? Origins = null,
    List<string>? LoreNotes = null
);

public record BaseStats(
    Dictionary<string, int> AbilityScoreIncrease,
    double SpeedM,
    string Size,
    double DarkvisionM,
    Height Height,
    Weight Weight,
    Age Age,
    Alignment Alignment,
    FlexibleAsi? FlexibleAsi = null,
    string? SpeedNotes = null,
    bool? SpeedNotReducedByHeavyArmor = null
);

public record Height(string Desc, string? RangeM = null);
public record Weight(string Desc, string? RangeKg = null);

public record Age(
    int MaturityYears,
    int LifespanYears,
    string Desc,
    int? AdulthoodCulturalYears = null,
    int? LifespanMaxYears = null
);

public record Alignment(string Tendency, string Desc);

public record FlexibleAsi(int Count, int Value, List<string> Excluded);

public record Trait(
    string Id,
    string Name,
    string Desc,
    JsonElement? Mechanics = null
);

public record CreationChoice(
    string Id,
    string Name,
    string Desc,
    string Type,
    int? ChoiceCount = null,
    JsonElement? Options = null,
    JsonElement? OptionGroups = null,
    string? SpellList = null,
    int? SpellLevel = null,
    string? SpellcastingAbility = null,
    int? ValuePerChoice = null,
    List<string>? Excluded = null
);

public record Subspecies(
    string Id,
    string Name,
    bool Playable,
    string Flavor,
    Dictionary<string, int> AbilityScoreIncrease,
    List<Trait> Traits,
    List<CreationChoice> CreationChoices,
    string? PlayableNotes = null,
    Languages? Languages = null
);

public record Languages(
    List<string> Fixed,
    int ChoiceCount,
    string? Notes = null,
    JsonElement? GrantsFromChoice = null
);

public record OptionalRule(
    string Id,
    string Name,
    string Desc,
    JsonElement? Mechanics = null
);

public record CivilizationLink(string Id, string Name, string Desc);